### PR TITLE
The current version of flux configuration 2022-03-01 doesn't add Postbuild substitution variables

### PR DESCRIPTION
**Pull request by aadev1**
# Description

<details>
<summary>The current version of flux configuration 2022-03-01 doesn't add Postbuild substitution variables</summary>
Updated Microsoft.KubernetesConfiguration/fluxConfigurations to version 2023-05-01
</details>

> Please include a summary of the change and which issue is fixed.

Updated Microsoft.KubernetesConfiguration/fluxConfigurations to version 2023-05-01 and added postBuild section in common  main.test.bicep.

> Please also include the context.

The current version 2022-03-01 doesn't add Postbuild substitution variables which are included in the Bicep configuration.  We have tested these by running the common/main.test.bicep with 2 postBuild substitute variables with values foo and bar and they don't appear in the Kustomization in AKS.  Please see snippet below which is a result of running 'kubectl describe kustomization fluxaakcfccom001-unified -n flux-system'

image

We have updated the API version to 2023-05-01 which renders the expected results.  Please see snippet below which is a result of running 'kubectl describe kustomization fluxaakcfccom001-unified -n flux-system'

image

This is a feature we require to correctly configure our Flux configuration.

> List any dependencies that are required for this change.

There are no dependencies

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

**Comment by aadev1** [comment]
@microsoft-github-policy-service agree

**Comment by aadev1** [comment]
@microsoft-github-policy-service agree company="defra"

**Comment by AlexanderSehr** [comment]
Hey @aadev1,

you may not have noticed, but we're currently in the process of migrating the CARML modules into the [Public Bicep Registry](https://github.com/Azure/bicep-registry-modules/tree/main/avm/res) repository. 
As the change you propose here is faily small, I'll port it over manually. But for future updates to any module with the [MOVED-TO-AVM.md](https://github.com/aadev1/ResourceModules/blob/users/aadev1/updateFluxConfigApuiVersion/modules/kubernetes-configuration/flux-configuration/MOVED-TO-AVM.md) file, I'd ask you to please contribute to that repository. :)

I know this may be a bit counter-intuitive right now, but we try to get through this transition period as fast as possible.

**Comment by aadev1** [comment]
Hi @AlexanderSehr,

Sorry I didn't realize you'd moved to AVM.  I'll make sure I contribute there for future changes.

I have tested this change using the common flux configuration test (modules\kubernetes-configuration\flux-configuration\.test\common\main.test.bicep) with postBuild.substitute variables set.

I've added screenshots to the description of results, using api version 2022-03-01 which doesn't five the desired results and api version 2023-05-01 which does give expected results.

**Comment by AlexanderSehr** [comment]
> Hi @AlexanderSehr,
> 
> Sorry I didn't realize you'd moved to AVM. I'll make sure I contribute there for future changes.
> 
> I have tested this change using the common flux configuration test (modules\kubernetes-configuration\flux-configuration.test\common\main.test.bicep) with postBuild.substitute variables set.
> 
> I've added screenshots to the description of results, using api version 2022-03-01 which doesn't five the desired results and api version 2023-05-01 which does give expected results.

All good @aadev1, this PR is the first case in the current period ;) I'll take care of the PR as soon as I can :)

File: modules/kubernetes-configuration/flux-configuration/main.bicep [hunk 1]
```diff
@@ -63,7 +63,7 @@ resource managedCluster 'Microsoft.ContainerService/managedClusters@2022-07-01'
   name: clusterName
 }
 
-resource fluxConfiguration 'Microsoft.KubernetesConfiguration/fluxConfigurations@2022-03-01' = {
+resource fluxConfiguration 'Microsoft.KubernetesConfiguration/fluxConfigurations@2023-05-01' = {
```
Comment (AlexanderSehr) [inline modules/kubernetes-configuration/flux-configuration/main.bicep:66]:
```suggestion
resource fluxConfiguration 'Microsoft.KubernetesConfiguration/fluxConfigurations@2022-11-01' = {
```

@aadev1, I successfully tested this API version which is newer than the old one, but not as new as the one you proposed. 
Note: This change would require you to regenerate the readme - which I guess you'd need to do anyways using the `utilities/tools/Set-ModuleReadMe.ps1` utility - and compile the JSON template using the `bicep build '<bicepPath>'` utility.

Comment (AlexanderSehr) [reply modules/kubernetes-configuration/flux-configuration/main.bicep:66]: Ok nevermind, different plan

File: modules/kubernetes-configuration/flux-configuration/main.bicep [hunk 2]
```diff
@@ -63,7 +63,7 @@ resource managedCluster 'Microsoft.ContainerService/managedClusters@2022-07-01'
   name: clusterName
```
Comment (AlexanderSehr) [inline modules/kubernetes-configuration/flux-configuration/main.bicep:63]: @JPEasier found out that with the new API version, the `kustomizations` become mandatory and that's why it is failing (even without an error message). I'll implement that change later in the AVM contribution, and the same should be done here in CAMRL.

Snippet:
```bicep
kustomizations: {
  unified: {
    dependsOn: []
    force: false
    path: './cluster-manifests'
    prune: false
    syncIntervalInSeconds: 300
    timeoutInSeconds: 300
  }
}
```

File: modules/kubernetes-configuration/flux-configuration/main.bicep (hunk 1 above)
Comment (JPEasier) [reply modules/kubernetes-configuration/flux-configuration/main.bicep:66]: Using Api version 2023-05-01 need at least one kustomizations. If you switch to the newest version please set this parameter to Required.

**Comment by eriqua** [comment]
@JPEasier @aadev1 @AlexanderSehr since this change has already neen merged to the public bicep registry, should we close this PR and mark the module as moved?

**Comment by AlexanderSehr** [comment]
> @JPEasier @aadev1 @AlexanderSehr since this change has already been merged to the public bicep registry, should we close this PR and mark the module as moved?

As @aadev1 is using CARML and not AVM I'll leave it to him. The original plan was to merge this PR to CARML regardless while I'd take care of the AVM part (which I did).

**Comment by aadev1** [comment]
@eriqua Could we leave it open for now please.  I'll look to make the 'kustomizations' mandatory as @AlexanderSehr has done in AVM.  Our organization is currently using CARML and we would like to get these changes merged.  If there's anything else that needs doing to get this PR merged, could you please let me know.

**Comment by AlexanderSehr** [comment]
> @eriqua Could we leave it open for now please. I'll look to make the 'kustomizations' mandatory as @AlexanderSehr has done in AVM. Our organization is currently using CARML and we would like to get these changes merged. If there's anything else that needs doing to get this PR merged, could you please let me know.

Hey @aadev1 not much, no. To get it merged we should get the changes implemented in [AVM](https://github.com/Azure/bicep-registry-modules/tree/main/avm/res/kubernetes-configuration/flux-configuration) in & resolve the current conflicts shown in this PR.

File: modules/kubernetes-configuration/flux-configuration/main.bicep (hunk 2 above)
Comment (aadev1) [reply modules/kubernetes-configuration/flux-configuration/main.bicep:63]: @AlexanderSehr @eriqua @JPEasier - I've updated the CARML repo to reflect the change made by @AlexanderSehr in AVM.  Could you please review.

**Review by AlexanderSehr — APPROVED** [review]

**Comment by AlexanderSehr** [comment]
Looks good. Thanks @aadev1 for putting in all the extra work. It's much appreciated

**Pull request merged**